### PR TITLE
Add support for parsing f-string as per PEP 701

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR adds support for PEP 701 in the parser to use the new tokens emitted by the lexer to construct the f-string node.

### Grammar

Without an official grammar, the f-strings were parsed manually. Now that we've the specification, that is being used in the LALRPOP to parse the f-strings.

### `string.rs`

This file includes the logic for parsing string literals and joining the implicit string concatenation. Now that we don't require parsing f-strings manually a lot of code involving the same is removed.

Earlier, there were 2 entry points to this module:
* `parse_string`: Used to parse a single string literal
* `parse_strings`: Used to parse strings which were implicitly concatenated

Now, there are 3 entry points:
* `parse_string_literal`: Renamed from `parse_string`
* `parse_fstring_middle`: Used to parse a `FStringMiddle` token which is basically a string literal without the quotes
* `concatenate_strings`: Renamed from `parse_strings` but now it takes the parsed nodes instead. So, we just need to concatenate them into a single node.

> A short primer on `FStringMiddle` token: This includes the portion of text inside the f-string that's not part of the expression and isn't an opening or closing brace. For example, in `f"foo {bar:.3f{x}} bar"`, the `foo `, `.3f` and ` bar` are `FStringMiddle` token content.

### `Constant::kind` changed in the AST

***Discussion in the official implementation: https://github.com/python/cpython/pull/102855#issuecomment-1508760963***

This change in the AST is when unicode strings (prefixed with `u`) and f-strings are used in an implicitly concatenated string value. For example,

```python
u"foo" f"{bar}" "baz" " some"
```

Pre Python 3.12, the kind field would be assigned only if the prefix was on the first string. So, taking the above example, both `"foo"` and `"baz some"` (implicit concatenation) would be given the `u` kind:

<details><summary>Pre 3.12 AST:</summary>
<p>

```python
Constant(value='foo', kind='u'),
FormattedValue(
  value=Name(id='bar', ctx=Load()),
  conversion=-1),
Constant(value='baz some', kind='u')
```

</p>
</details> 

But, post Python 3.12, only the string with the `u` prefix will be assigned the value:

<details><summary>Pre 3.12 AST:</summary>
<p>

```python
Constant(value='foo', kind='u'),
FormattedValue(
  value=Name(id='bar', ctx=Load()),
  conversion=-1),
Constant(value='baz some')
```

</p>
</details>

Here are some more iterations around the change:

1. `"foo" f"{bar}" u"baz" "no"`

<details><summary>Pre 3.12</summary>
<p>

```python
Constant(value='foo'),
FormattedValue(
  value=Name(id='bar', ctx=Load()),
  conversion=-1),
Constant(value='bazno')
```

</p>
</details>

<details><summary>3.12</summary>
<p>

```python
Constant(value='foo'),
FormattedValue(
  value=Name(id='bar', ctx=Load()),
  conversion=-1),
Constant(value='bazno', kind='u')
```

</p>
</details> 

2. `"foo" f"{bar}" "baz" u"no"`

<details><summary>Pre 3.12</summary>
<p>

```python
Constant(value='foo'),
FormattedValue(
  value=Name(id='bar', ctx=Load()),
  conversion=-1),
Constant(value='bazno')
```

</p>
</details>

<details><summary>3.12</summary>
<p>

```python
Constant(value='foo'),
FormattedValue(
  value=Name(id='bar', ctx=Load()),
  conversion=-1),
Constant(value='bazno')
```

</p>
</details>

3. `u"foo" f"bar {baz} realy" u"bar" "no"`

<details><summary>Pre 3.12</summary>
<p>

```python
Constant(value='foobar ', kind='u'),
FormattedValue(
  value=Name(id='baz', ctx=Load()),
  conversion=-1),
Constant(value=' realybarno', kind='u')
```

</p>
</details>

<details><summary>3.12</summary>
<p>

```python
Constant(value='foobar ', kind='u'),
FormattedValue(
  value=Name(id='baz', ctx=Load()),
  conversion=-1),
Constant(value=' realybarno')
```

</p>
</details> 

### Errors

With the hand written parser, we were able to provide better error messages in case of any errors such as the following but now they all are removed and in those cases an "unexpected token" error will be thrown by lalrpop:
* A closing delimiter was not opened properly
* An opening delimiter was not closed properly
* Empty expression not allowed

The "Too many nested expressions in an f-string" was removed and instead we can create a lint rule for that.

And, "The f-string expression cannot include the given character" was removed because f-strings now support those characters which are mainly same quotes as the outer ones, escape sequences, comments, etc.

## Test Plan

1. Refactor existing test cases to use `parse_suite` instead of `parse_fstrings` (doesn't exists anymore)
2. Additional test cases are added as required

Updated the snapshots. The change from `parse_fstrings` to `parse_suite` means that the snapshot would produce the module node instead of just a list of f-string parts. I've manually verified that the parts are still the same along with the node ranges.

## Benchmarks

https://github.com/astral-sh/ruff/pull/7263#issue-1889732755

fixes: #7043
fixes: #6835

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #7035** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7035" target="_blank">Graphite</a> 
    * **PR #7013** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7013" target="_blank">Graphite</a> 
      * **PR #6659** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6659" target="_blank">Graphite</a> 
        * **PR #7041** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7041" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/7041?utm_source=stack-comment).

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:297]: Nit: this requires two full iterations. We can reduce it to one by using a plain loop

```rust
let mut has_fstring = false;
let mut byte_literal_count = 0u32;

for string in strings {
    has_fstring |= pexpr.is_f_string_expr();

    if is_bytes {
        byte_literal_count += 1;
    }
}
```

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:277]: What's the reason for passing `ParenthesizedExpr`? This should always either be StringLiteral, ByteLiteral, or FStringExpr nodes right? I recommend creating a new enum with just these three variants. It's very likely that that new union is smaller in size which helps performance.

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:178]: Lol, we should change kind to an enum. Allocating a string for this is silly (and I'm surprised the kind is a flag on constant when only strings can be unicode). But we can do this as a separate PR.

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:259]: Could you expand this comment with an example and why it is necessary? I know we talked about it on Discord but new and external maintainers won't know about the conversation or don't have access to it.

Overall: this is unfortunate because it requires moving all parts into a new Vec (flatten, then collect). I wonder if a dedicated token would be a better solution

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:320]: I wonder if we should add a fast path (for the whole method) that handles the most common case of a single string:

* No need to test if it is a mixed byte / non-byte concatenation
* No need to copy the value (value.into_bytes() or use the original string)

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:333]: We should change this to a `String` and use `push_str` rather than building a vec first and than joining the parts.

File: crates/ruff_python_parser/src/string.rs
Comment (MichaReiser) [inline crates/ruff_python_parser/src/string.rs:367]: Should we change this to a `String` and use `push_str` directly instead of pushing parts and joining them at the end?

**Review by MichaReiser — APPROVED** [review]
Nice work! Overall this is looking good to me. It will be interesting to get some ecosystem checks once ruff compiles. 

Most of my comments are nits but I think there's potential to clean up `string.rs` further (and hopefully improving performance at the same time). I leave it up to you if you want to do this as part of this or a follow up PR.

File: crates/ruff_python_parser/src/string.rs
Comment (konstin) [inline crates/ruff_python_parser/src/string.rs:2]: nit:
```suggestion
//! Parsing of string literals, byte literals and implicit string concatenation.
```

**Review by konstin — APPROVED** [review]
The infinite loop for something that succesfully fails in the unit test is strange

File: crates/ruff/src/linter.rs
Comment (dhruvmanila) [reply crates/ruff/src/linter.rs:151]: Opened #7153 :)

File: crates/ruff_python_parser/src/python.lalrpop
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/python.lalrpop:1622]: I'm unable to follow your comment with the highlighted code. The `None` is assigned to `debug_text` via the following condition which now that I'm looking at can be refactored using `debug.map(|_| {...})`:
```rust
        let debug_text = if debug.is_some() {
            let start_offset = location + TextSize::from(1);  // +1 to skip the opening brace
            let end_offset = if let Some((conversion_start, conversion_flag)) = fstring_conversion {
                conversion = conversion_flag;
                conversion_start
            } else {
                // Subtract 1 to account for the closing brace or the `:` in the format spec
                format_spec.as_ref().map_or(end_location, |spec| spec.range().start()) - TextSize::from(1)
            };
            Some(ast::DebugText {
                leading: source_code[TextRange::new(start_offset, value.range().start())].to_string(),
                trailing: source_code[TextRange::new(value.range().end(), end_offset)].to_string(),
            })
        } else {
            None
        };
```

File: crates/ruff_python_parser/src/python.lalrpop
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/python.lalrpop:1635]: It isn't included in the official CPython implementation nor our earlier implementation.

https://play.ruff.rs/f1927a7d-bbb2-4afb-b56c-fffdaa74fa44

```
>>> mod.body[0].value.values[0].format_spec.values[0].col_offset
7
>>> mod.body[0].value.values[0].format_spec.values[0].end_col_offset
10
>>> s = 'f"{foo:.3f}"'
>>> s[7:10]
'.3f'
```

File: crates/ruff_python_parser/src/string.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/string.rs:277]: Yeah I actually thought of using that but it'll require moving the kind field (for unicode) into `StringConstant`. Giving a quick look at the usage of the kind field, it does not seem to be used many places so should be a quick refactor.

File: crates/ruff_python_parser/src/string.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/string.rs:178]: I think it should be a boolean field `unicode` similar to `implicit_concatenated` on `StringConstant`. What do you think?

File: crates/ruff_python_parser/src/python.lalrpop
Comment (MichaReiser) [reply crates/ruff_python_parser/src/python.lalrpop:1622]: What I understand from the code:

* The `conversion` field is always `ConversionFlag::None` if `debug` is `None`
* But the grammar accepts `{test!s}` 

The grammer is correct in my view but the initialization of the `conversion` field is missing.

File: crates/ruff_python_parser/src/python.lalrpop
Comment (MichaReiser) [reply crates/ruff_python_parser/src/python.lalrpop:1635]: I would then lean towards moving the `:` into the `FStringReplacementField` to make it clear that the `:` is a separator and isn't part of the format spec node.

File: crates/ruff_python_parser/src/python.lalrpop
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/python.lalrpop:1622]: Oh right, that's a good find. I missed the initialization part. Thanks for noticing that!

File: crates/ruff_python_parser/src/string.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/string.rs:367]: Yes, we can then use `std::mem::take` to move the `String`.

File: crates/ruff_python_parser/src/python.lalrpop
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/python.lalrpop:1635]: I'm using a similar pattern as `AssignSuffix`:

```
FStringFormatSpecSuffix = {
	":" <FStringFormatSpec>
}

FStringFormatSpec = {
	<location:@L> ...
}
```

File: crates/ruff_python_parser/src/parser.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/parser.rs:1278]: Yeah, I'm still a bit on the edge with the last test case in this function :/

File: crates/ruff_python_parser/src/lexer.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/lexer.rs:2097]: This is a good find. The intention was not to check for infinite loop. The test function stops when it encounters an error but the lexer goes into an infinite loop when it's otherwise used.

File: crates/ruff_python_parser/src/string.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/string.rs:298]: PR: #7211

File: crates/ruff_python_parser/src/lexer.rs
Comment (dhruvmanila) [reply crates/ruff_python_parser/src/lexer.rs]: Also, note that `cargo run --bin ruff` won't work right now as the linter changes w.r.t. the new f-string tokens are yet to be implemented.

**Comment by hugovk** [comment]
Thanks for this PR!

I've tested it against CPython's f-string test file, and it runs as expected:

https://github.com/python/cpython/blob/3.12/Lib/test/test_fstring.py

Whereas `main` unsurprisingly fails to parse:

```
error: Failed to parse /Users/hugo/github/cpython/Lib/test/test_fstring.py:671:28: f-string: expecting '}'
/Users/hugo/github/cpython/Lib/test/test_fstring.py:671:28: E999 SyntaxError: f-string: expecting '}'
```

It would be great if this was ready by the time Python 3.12 is released on 2023-10-02 ([PEP 693](https://peps.python.org/pep-0693/)).

**Comment by dhruvmanila** [comment]
> Thanks for this PR!

Hey, thanks for the comment. We're in the final phase of making the necessary changes across other parts of the codebase (linter and formatter) and then we'll merge this.

**Pull request merged**
